Training/interface/DataLoader_main.h

```diff
@@ -0,0 +1,931 @@
+      {
+          static constexpr double iso_cone = 0.5;
+
+          CellGrid grid = cellGridRef;
+          const double tau_pt = tau.tau_pt, tau_eta = tau.tau_eta, tau_phi = tau.tau_phi;
+
+          const auto fillGrid = [&](CellObjectType type, const std::vector<float>& eta_vec,
+                                    const std::vector<float>& phi_vec, const std::vector<int>& pdgId = {}) {
+              if(eta_vec.size() != phi_vec.size())
+                  throw std::runtime_error("Inconsistent cell inputs.");
+              for(size_t n = 0; n < eta_vec.size(); ++n) {
+                  if(pdgId.size() && GetCellObjectType(pdgId.at(n)) != type) continue;
+                  const double eta = eta_vec.at(n), phi = phi_vec.at(n);
+                  const double deta = eta - tau_eta, dphi = DeltaPhi(phi, tau_phi);
+                  const double dR = std::hypot(deta, dphi);
+                  const bool inside_signal_cone = dR < getInnerSignalConeRadius(tau_pt);
+                  const bool inside_iso_cone = dR < iso_cone;
+                  if(inner && !inside_signal_cone) continue;
+                  // if(!inner && (inside_signal_cone || !inside_iso_cone)) continue;
+                  if(!inner && !inside_iso_cone) continue;
+                  CellIndex cellIndex;
+                  if(grid.TryGetCellIndex(deta, dphi, cellIndex))
+                      grid.at(cellIndex)[type].insert(n);
+              }
+          };
+
+          fillGrid(CellObjectType::PfCand_electron, tau.pfCand_eta, tau.pfCand_phi, tau.pfCand_pdgId);
+          fillGrid(CellObjectType::PfCand_muon, tau.pfCand_eta, tau.pfCand_phi, tau.pfCand_pdgId);
+          fillGrid(CellObjectType::PfCand_chargedHadron, tau.pfCand_eta, tau.pfCand_phi, tau.pfCand_pdgId);
+          fillGrid(CellObjectType::PfCand_neutralHadron, tau.pfCand_eta, tau.pfCand_phi, tau.pfCand_pdgId);
+          fillGrid(CellObjectType::PfCand_gamma, tau.pfCand_eta, tau.pfCand_phi, tau.pfCand_pdgId);
+          fillGrid(CellObjectType::Electron, tau.ele_eta, tau.ele_phi);
+          fillGrid(CellObjectType::Muon, tau.muon_eta, tau.muon_phi);
+
+          return grid;
+      }
+
+private:
+  const size_t n_tau; // number of events(=taus)
+  Long64_t start_dataset;
+  Long64_t end_dataset;
+  Long64_t current_entry; // number of the current entry
+  const size_t n_inner_cells;
+  const double inner_cell_size;
+  const size_t n_outer_cells;
+  const double outer_cell_size;
+  const int n_threads;
+  const size_t n_fe_tau;
+  const size_t n_pf_el;
+  const size_t n_pf_mu;
+  const size_t n_pf_chHad;
+  const size_t n_pf_nHad;
+  const size_t n_pf_gamma;
+  const size_t n_electrons;
+  const size_t n_muons;
+  const int n_labels;
+  const std::vector<std::string>  labels_names;
+  const int n_eta_bins;
+  const double eta_min;
+  const double eta_max;
+  const int n_pt_bins;
+  const double pt_min;
+  const double pt_max;
```

Comment:
these variables are already defined in setup, right? If yes, I don't see the necessity to repeat them here